Issue 47655: Support navigation keys for editable grid dropdown cells

labkey-nicka opened the pull request:
#### Rationale
This work seeks to address [Issue 47655](https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=47655) with the following changes to the Editable Grid:

1. Dropdown cells now support `TAB` and `ESCAPE` key handling to better support grid navigation.
1. Dropdown cells will always `openMenuOnFocus`.

In doing so the dropdown cells should immediately open upon clicking on the dropdown indicator.

#### Related Pull Requests
- https://github.com/LabKey/labkey-ui-components/pull/1225
- https://github.com/LabKey/labkey-ui-premium/pull/124
- https://github.com/LabKey/biologics/pull/2213
- https://github.com/LabKey/sampleManagement/pull/1931
- https://github.com/LabKey/inventory/pull/910

#### Changes
- Add `onKeyDown` event handling for `TAB` and `ESCAPE` for dropdown cells.
- Expose `onKeyDown` event handling prop on `SelectInput`.
- Introduce `showDropdownIndicator`, `showIndicatorSeparator` and `showDropdownMenu` render bits on `SelectInput`.
- Default dropdown cells to `showIndicatorSeparator=false` to better match editable grid styling.
- Fix round-tripping values within the "Alias" column on an editable grid via an `AliasGridInput` processing component.
- Introduce types of `CellCoordinates` and `GridColumnCellRenderer`.
- Update `SampleStatusInput` handling of `containerFilter` to use prop override pattern instead of `delete`.

labkey-nicka committed:
SelectInput: add onKeyDown, showDropdownIndicator, showDropdownMenu, showIndicatorSeparator

labkey-nicka committed:
LookupCell: Support onKeyDown, always openMenuOnFocus
- multi-select case tabbing is supported via onKeyDown

labkey-nicka committed:
Cell: Support Tab, Escape actions from LookupCells

labkey-nicka committed:
LookupCell: default to showIndicatorSeparator=false
- better matches editable grid cell styling

labkey-nicka committed:
showIndicatorSeparator default to true

labkey-nicka committed:
Support Key.TAB

labkey-nicka committed:
AliasGridInput: fix value round-trip within editable grid

labkey-nicka committed:
v2.349.3-fb-edit-grid-47655.0

labkey-nicka committed:
CellCoordinates

labkey-nicka committed:
GridColumnCellRenderer

labkey-nicka committed:
v2.349.3-fb-edit-grid-47655.1

labkey-nicka committed:
Test updates

labkey-nicka committed:
v2.349.3-fb-edit-grid-47655.2

cnathe commented on packages/components/src/internal/components/samples/DiscardConsumedSamplesPanel.tsx:
```diff
@@ -28,7 +28,7 @@ export const DiscardConsumedSamplesPanel: FC<Props> = memo(props => {
                                 checked={shouldDiscard}
                             />
-                            <span className="left-spacing">{discardTitle}</span>
+                            <span className="discard-consumed-title left-spacing">{discardTitle}</span>
```
I have a feeling we will be doing a lot of this, i.e. adding a className, for the enzyme to RTL conversion.

cnathe reviewed (approved):
Looks good. I did some manual testing as well and the fix addresses the original issue mentioned in 47655 and the interactions for the tab key work great.

Note that since we've branched for 23.7, if this ends up being the first ui-components PR merged since then, make sure to bump the minor version to leave room for 23.7 patches. Thanks.

labkey-nicka committed:
lint

labkey-nicka committed:
SelectInput: expose closeMenuOnSelect and tabSelectsValue
- ensure menu closes on menu item selection when closeMenuOnSelect={true} and isMulti={true}

labkey-nicka committed:
v2.349.3-fb-edit-grid-47655.3

labkey-nicka committed:
SelectInput: closeMenuOnSelect, tabSelectsValue default to true
- same as ReactSelect

labkey-nicka committed:
v2.349.3-fb-edit-grid-47655.4

labkey-nicka committed:
SelectInput: restrict check to when openMenuOnFocus={true}

labkey-nicka committed:
v2.349.3-fb-edit-grid-47655.5

labkey-nicka committed:
Prepare release notes

labkey-nicka committed:
v2.350.0
